bpo-41894: Fix UTF-8 errors loading native module

kadler opened the pull request:
When running in a non-UTF-8 locale, if an error occurs while importing a
native Python module (say because a dependent share library is missing),
the error message string returned may contain non-ASCII code points
causing a UnicodeDecodeError.

While this is most likely to occur on AIX, which still uses non-UTF-8
locales by default, it could occur on any Unix-like system.

<!--
Thanks for your contribution!
Please read this comment in its entirety. It's quite important.

# Pull Request title

It should be in the following format:

```
bpo-NNNN: Summary of the changes made
```

Where: bpo-NNNN refers to the issue number in the https://bugs.python.org.

Most PRs will require an issue number. Trivial changes, like fixing a typo, do not need an issue.

# Backport Pull Request title

If this is a backport PR (PR made against branches other than `master`),
please ensure that the PR title is in the following format:

```
[X.Y] <title from the original PR> (GH-NNNN)
```

Where: [X.Y] is the branch name, e.g. [3.6].

GH-NNNN refers to the PR number from `master`.

-->


<!-- issue-number: [bpo-41894](https://bugs.python.org/issue41894) -->
https://bugs.python.org/issue41894
<!-- /issue-number -->

kadler committed:
bpo-41894: Fix UTF-8 errors loading native module

When running in a non-UTF-8 locale, if an error occurs while importing a
native Python module (say because a dependent share library is missing),
the error message string returned may contain non-ASCII code points
causing a UnicodeDecodeError.

PyUnicode_DecodeFSDefault is used for buffers which may contain
filesystem  paths. For consistency with os.strerror(),
PyUnicode_DecodeLocale is used for buffers which contain system error
messages. While the shortname parameter is always encoded in ASCII
according to PEP 489, it is left decoded using PyUnicode_FromString to
minimize the changes and since it should not affect the decoding (albeit
_potentially_ slower).

In dynload_hpux, since the error buffer contains a message generated
from a static ASCII string and the module filesystem path,
PyUnicode_DecodeFSDefault is used instead of PyUnicode_DecodeLocale as
is used elsewhere.

methane commented on Python/dynload_shlib.c:
```diff
@@ -106,7 +106,7 @@ _PyImport_FindSharedFuncptr(const char *prefix,
         if (error == NULL)
             error = "unknown dlopen() error";
-        error_ob = PyUnicode_FromString(error);
+        error_ob = PyUnicode_DecodeLocale(error, "surrogateescape");
```
Can we use `PyUnicode_DecodeFSDefault` instead?
Error message may contain file path.

ZackerySpytz commented on Python/dynload_aix.c:
```diff
@@ -145,8 +145,8 @@ aix_loaderror(const char *pathname)
+    pathname_ob = PyUnicode_DecodeFSDefault(pathname);
+    errbuf_ob = PyUnicode_DecodeFSDefault(errbuf);
     PyErr_SetImportError(errbuf_ob, NULL, pathname);
     Py_DECREF(pathname_ob);
     Py_DECREF(errbuf_ob);
```
These ``Py_DECREF()`` calls will cause segfaults if ``PyUnicode_DecodeFSDefault()`` fails.

kadler replied:
I've added NULL checks.

ZackerySpytz commented on Python/dynload_aix.c:
```diff
@@ -145,8 +145,8 @@ aix_loaderror(const char *pathname)
-    errbuf_ob = PyUnicode_FromString(errbuf);
+    pathname_ob = PyUnicode_DecodeFSDefault(pathname);
+    errbuf_ob = PyUnicode_DecodeFSDefault(errbuf);
     PyErr_SetImportError(errbuf_ob, NULL, pathname);
```
It seems that ``pathname_ob`` should be passed here, not ``pathname``.

kadler replied:
Good catch, fixed.

kadler committed:
bpo-41894: Fix bugs in dynload error msg handling

For both dynload_aix and dynload_hpux, properly handle the possibility
that decoding strings may return NULL and when such an error happens,
properly decrement any previously decoded strings and return early.

In addition, in dynload_aix, ensure that we pass the decoded string
*object* pathname_ob to PyErr_SetImportError instead of the original
pathname buffer.

Co-authored-by: Serhiy Storchaka <storchaka@gmail.com>

serhiy-storchaka commented on Python/dynload_aix.c:
```diff
@@ -144,12 +144,13 @@ aix_loaderror(const char *pathname)
-    Py_DECREF(errbuf_ob);
+    /* Subtract 1 from the length to trim off trailing newline */
+    errbuf_ob = PyUnicode_DecodeFSDefaultAndSize(errbuf, strlen(errbuf)-1);
+    pathname_ob = PyUnicode_DecodeFSDefault(pathname);
```
`PyErr_SetImportError()` can fail if `errbuf_ob` or `pathname_ob` are NULL.

```suggestion
    errbuf_ob = PyUnicode_DecodeFSDefaultAndSize(errbuf, strlen(errbuf)-1);
    if (errbuf_ob == NULL) {
        return;
    }
    pathname_ob = PyUnicode_DecodeFSDefault(pathname);
    if (pathname_ob == NULL) {
        Py_DECREF(errbuf_ob);
        return;
    }
```

kadler replied:
Since we now check for NULL prior to calling PyErr_SetImportError, I remove the NULL checks I had previously added before calling `Py_DECREF` on the two objects.

serhiy-storchaka commented on Misc/NEWS.d/next/Core and Builtins/2020-10-02-11-35-33.bpo-41894.ffmtOt.rst:
```diff
@@ -0,0 +1,4 @@
+When loading a native module and a load failure occurs, prevent a possible
+UnicodeDecodeError when not running in a UTF-8 locale by decoding the load
+error message with ``PyUnicode_DecodeFSDefault`` instead of
```
I think it is better to not mentioned PyUnicode_DecodeFSDefault and PyUnicode_FromString (which are implementation details and don't say much to common user) but just say that locale encoding is used instead of UTF-8.

serhiy-storchaka reviewed (approved):
LGTM

serhiy-storchaka commented on Python/dynload_aix.c:
```diff
@@ -144,10 +144,18 @@ aix_loaderror(const char *pathname)
+    if (errbuf_ob == NULL) {
+        return;
+    }
+    pathname_ob = PyUnicode_DecodeLocale(pathname, "surrogateescape");
```
For `pathname` we use `PyUnicode_EncodeFSDefault()` in `_PyImport_LoadDynamicModuleWithSpec()`, so I think we should use `PyUnicode_DecodeFSDefault()` here.

And please use the FS encoding  for pathname and locale encoding for system error message also in `dynload_shlib.c` and `/dynload_hpux.c`.

kadler replied:
Ok, so error messages use PyUnicode_DecodeLocale w/ surrogateescape but for filesystem paths use PyUnicode_DecodeFSDefault. What about shortname, eg. https://github.com/python/cpython/blob/master/Python/dynload_shlib.c#L112 ?

kadler replied:
For dynload_hpux, it seems that the error message is composed of an ASCII-only literal string and the path: https://github.com/python/cpython/blob/master/Python/dynload_hpux.c#L37-L39 so it would seem that PyUnicode_DecodeFSDefault would be best here, no?

kadler replied:
Answering my own question about encoding of shortname. According to PEP 489, it will always be encoded in ASCII: https://www.python.org/dev/peps/pep-0489/#export-hook-name - either directly or by being encoded with "punycode".

serhiy-storchaka reviewed (approved):
LGTM.
